src/theory/theory_model.h

```diff
@@ -130,12 +130,26 @@ class TheoryModel : public Model
   void printRepresentativeDebug( const char* c, Node r );
   /** print representative function */
   void printRepresentative( std::ostream& out, Node r );
+private:
+  /** uf models */
+  std::map< Node, Node > d_uf_models;
 public:
   /** whether function models are enabled */
   bool d_enableFuncModels;
-  //necessary information for function models
+  /** a map from functions f to a list of all APPLY_UF terms with operator f */
   std::map< Node, std::vector< Node > > d_uf_terms;
-  std::map< Node, Node > d_uf_models;
+  /** a map from functions f to a list of all HO_APPLY terms with first argument f */
+  std::map< Node, std::vector< Node > > d_ho_uf_terms;
+  /** assign function value f to definition f_def */
+  void assignFunctionDefinition( Node f, Node f_def );
+  /** have we assigned function f? */
+  bool hasAssignedFunctionDefinition( Node f ) const { return d_uf_models.find( f )!=d_uf_models.end(); }
+  /** get the list of functions to assign. 
```

Comment:
Please be more specific.  What are the criteria for the functions in the list and how is the list obtained?